lib/reports/by_country_report.rb

```diff
@@ -0,0 +1,94 @@
+require 'net/http'
+
+class ByCountryReport
+  def initialize
+    @registered_firms = Firm.registered
+    @lookup = fetch_countries
+  end
+
+  def firm_report
+    firms = @registered_firms.where(parent_id: nil) # don't include trading names
+    grouped_firms = firms.group_by do |firm|
+      @lookup[postcode_slug_for(firm)]
+    end
+    generate_report('Firms', grouped_firms, firms.count)
+  end
+
+  def adviser_report
+    advisers = @registered_firms.map(&:advisers).flatten # include trading names
+    grouped_advisers = advisers.group_by do |adviser|
+      @lookup[postcode_slug_for(adviser.firm)]
+    end
+    generate_report('Advisers', grouped_advisers, advisers.count)
+  end
+
+  def error_report
+    grouped_firms = @registered_firms.group_by do |firm|
+      @lookup[postcode_slug_for(firm)]
+    end
+    firms_with_errors = grouped_firms.select { |k, _| k.blank? }.values.flatten.map(&:id).uniq
```

Comment:
`k` will be blank for any postcodes which return a blank country (i.e. those that are not found on the postcode lookup).